client/asset/dcr/dcr.go

```diff
@@ -2195,64 +2179,85 @@ func (dcr *ExchangeWallet) signTx(baseTx *wire.MsgTx) (*wire.MsgTx, error) {
-	if subtractee == nil {
-		subtractee = changeOutput
-		changeFees := dexdcr.P2PKHOutputSize * feeRate
-		if changeFees+minFee > remaining { // Catch underflow
-			changeOutput.Value = 0
+	totalIn, totalOut, remaining, _, size := reduceMsgTx(msgTx)
+	if totalIn < totalOut {
+		return nil, nil, "", 0, fmt.Errorf("unbalanced transaction")
+	}
+
+	minFee := feeRate * size
+	if subtractFrom == -1 && minFee > remaining {
+		return nil, nil, "", 0, fmt.Errorf("not enough funds to cover minimum fee rate. %.8f < %.8f",
+			toDCR(minFee), toDCR(remaining))
+	}
+	if int(subtractFrom) >= len(baseTx.TxOut) {
+		return nil, nil, "", 0, fmt.Errorf("invalid subtractFrom output %d for tx with %d outputs",
+			subtractFrom, len(baseTx.TxOut))
+	}
+
+	// Add a change output if there is enough remaining.
+	var changeAdded bool
+	var changeAddress dcrutil.Address
+	var changeOutput *wire.TxOut
+	minFeeWithChange := (size + dexdcr.P2PKHOutputSize) * feeRate
+	if remaining > minFeeWithChange {
+		changeValue := remaining - minFeeWithChange
+		if subtractFrom >= 0 {
+			// Subtract the additional fee needed for the added change output
+			// from the specified existing output.
+			changeValue = remaining
+		}
+		if !dexdcr.IsDustVal(dexdcr.P2PKHOutputSize, changeValue, feeRate) {
+			if subtractFrom >= 0 { // only subtract after dust check
+				baseTx.TxOut[subtractFrom].Value -= int64(minFeeWithChange)
+				remaining += minFeeWithChange
+			}
+			changeOutput, changeAddress, err = dcr.makeChangeOut(changeValue)
+			if err != nil {
+				return nil, nil, "", 0, err
+			}
+			dcr.log.Debugf("Change output size = %d, addr = %s", changeOutput.SerializeSize(), changeAddress.String())
+			changeAdded = true
+			baseTx.AddTxOut(changeOutput) // unsigned txn
+			remaining -= changeValue
 		} else {
-			changeOutput.Value -= int64(minFee + changeFees)
+			fmt.Println("DUSTING IT", changeValue, remaining, minFeeWithChange)
 		}
-	} else {
-		reservoir = uint64(subtractee.Value)
-	}
-
-	if minFee > reservoir {
-		return nil, nil, 0, fmt.Errorf("not enough funds to cover minimum fee rate. %.8f < %.8f",
-			toDCR(minFee), toDCR(reservoir))
 	}
 
-	// If the change is not dust, recompute the signed txn size and iterate on
-	// the fees vs. change amount.
-	changeAdded := !dexdcr.IsDust(subtractee, feeRate)
-	if changeAdded {
-		// Add the change output.
-		size0 := baseTx.SerializeSize()
-		baseTx.AddTxOut(changeOutput)
-		changeSize := baseTx.SerializeSize() - size0 // may be dexdcr.P2PKHOutputSize
-		dcr.log.Debugf("Change output size = %d, addr = %s", changeSize, addr.String())
+	lastFee := remaining
 
-		size += changeSize
-		lastFee = feeRate * uint64(size)
-		subtractee.Value = int64(reservoir - lastFee)
+	// If change added or subtracting from an existing output, iterate on fees.
+	if changeAdded || subtractFrom >= 0 {
+		subtractee := changeOutput
+		if subtractFrom >= 0 {
+			subtractee = baseTx.TxOut[subtractFrom]
+		}
+		reservoir := remaining + uint64(subtractee.Value)
```

Comment:
will comment on this in the code, but this reasoning is that `reservoir` is what is available for fees, `remaining` is what is presently already free and thus would be fees, and `subtractee.Value` is additional amount that can be add to fees or absorb excess fees from `remaining`.